runtime-sdk/src/module.rs

```diff
@@ -445,6 +445,46 @@ impl InvariantHandler for Tuple {
     }
 }
 
+/// Info handler.
+pub trait ModuleInfoHandler {
+    /// Reports info about the module(s).
+    fn module_info<C: Context>(_ctx: &mut C) -> BTreeMap<String, modules::core::types::ModuleInfo>;
+
+    /// Lists the names of all RPC methods exposed by this module. The result is informational
+    /// only; a return value of None means that the implementor does not care to list the methods,
+    /// or the implementor is a tuple of modules.
+    fn list_methods() -> Option<Vec<String>> {
```

Comment:
As a first/naive/lazy implementation, I was going to allow modules to override this method _if_ they feel like providing method introspection. But that doesn't work because (stable) Rust does not support method overrides.

Options:
- Do not provide method introspection.
- Go with current implementation, add `#![feature(specialization)]` to crate attributes to allow overrides. (I don't like this, it's kludgy anyway because it doesn't enforce the result list_methods() to be in sync with reality.)
- Implement macros that would require one to annotate call handlers like so:
  ```
  #[derive(MethodHandler)]
  pub struct Module {
    #[query_handler("core.MinGasPrice")]
    fn query_min_gas_price();

    #[call_handler("core.FooCall")]
    fn call_foo();
  }
  ```
  The `derive()` would implement `dispatch_query()` and `dispatch_call()` from `MethodHandler`, plus a new `list_methods()` fn.
- Similar but without macro magic: Add fields `QUERIES` and `CALLS` of type `BTreeMap<String, FnOnce>` to the `Module` signature.
  They describe the dispatch rules. `dispatch_query()`, `dispatch_call()`, and `list_methods()` are then implemented automatically in the default impl based on those two fields.
  
I would like to go with the last (or first) option but IDK if it's OK to change all the existing runtimes.

Reply:
I would be in favor of the macro variant.

Reply:
Good point re/ types; implemented with macros in #774